pkg/actions/karpenter/create.go

```diff
@@ -21,24 +21,33 @@ func (i *Installer) Create() error {
 	if err != nil {
 		return fmt.Errorf("unexpected or invalid ARN: %q, %w", i.Config.Status.ARN, err)
 	}
-	roleArn := fmt.Sprintf("arn:aws:iam::%s:policy/%s-%s", parsedARN.AccountID, builder.KarpenterManagedPolicy, i.Config.Metadata.Name)
 	clientSetGetter := &kubernetes.CallbackClientSet{
 		Callback: func() (kubernetes.Interface, error) {
 			return i.ClientSet, nil
 		},
 	}
-	karpenterServiceAccountTaskTree := i.StackManager.NewTasksToCreateIAMServiceAccounts([]*api.ClusterIAMServiceAccount{
-		{
-			ClusterIAMMeta: api.ClusterIAMMeta{
-				Name:      karpenter.DefaultServiceAccountName,
-				Namespace: karpenter.DefaultNamespace,
-			},
-			AttachRoleARN: roleArn,
+	var roleARN string
+	iamServiceAccount := &api.ClusterIAMServiceAccount{
+		ClusterIAMMeta: api.ClusterIAMMeta{
+			Name:      karpenter.DefaultServiceAccountName,
+			Namespace: karpenter.DefaultNamespace,
 		},
-	}, nil, clientSetGetter)
+	}
+	if api.IsDisabled(i.Config.Karpenter.CreateServiceAccount) {
+		policyArn := fmt.Sprintf("arn:aws:iam::%s:policy/%s-%s", parsedARN.AccountID, builder.KarpenterManagedPolicy, i.Config.Metadata.Name)
+		iamServiceAccount.AttachPolicyARNs = []string{policyArn}
+	} else {
+		// Create the service account role only.
+		roleName := fmt.Sprintf("eksctl-%s-iamservice-role", i.Config.Metadata.Name)
+		roleARN = fmt.Sprintf("arn:aws:iam::%s:role/%s", parsedARN.AccountID, roleName)
+		iamServiceAccount.RoleOnly = api.Enabled()
+		iamServiceAccount.RoleName = roleName
+		iamServiceAccount.AttachPolicy = makePolicyDocument()
```

Comment:
why does one scenario use `iamServiceAccount.AttachPolicyARNs = []string{policyArn}` and the other `iamServiceAccount.AttachPolicy = makePolicyDocument()` :thinking: shouldn't it be the same policies everytime?

Reply:
That's a good point. I guess we could use it since it has all the permissions and I could throw out the makePolicy thing. It was using it before I realised that I'm attaching a role when I should be attaching a policy instead.

But now, I'm sort of in a conundrum, where the policy is required, but it's actually only created after I'm getting to this point. And now I'm in a trap, because I can't create the Karpenter stuff, unless the policies exist. But I'm installing the Karpenter thing via a task separatly because I want a stack and now I have to refactor the whole thing and take out the Helm install part from the cloudformation stack part.

So create the CF stack with the roles and policies, add the mapping, add the accoutn, and run the helm install. I don't remember why I did it like this in the first place, but I do remember getting an error that something didn't exist and it needs to happen before that. I even have a comment about that somewhere. :D

Reply:
I found the comment, but it's so odd. I must have messed something up. Anyways, I understand things better now, so I'll refactor around it.